### PR TITLE
Make deno multithreaded

### PR DESCRIPTION
**Pull request by ry**

### CODE REVIEW

File: src/isolate.rs [hunk 1]
```diff
@@ -68,6 +144,38 @@ impl Isolate {
...
+  pub fn send(&self, buf: Buf) {
+    unsafe { libdeno::deno_send(self.ptr, buf.into()) };
+  }
+
+  // TODO Use Park abstraction? Note at time of writing Tokio default runtime
+  // does not have new_with_park().
+  pub fn event_loop(&self) {
+    // Main thread event loop.
+    loop {
+      match self.rx.try_recv() {
```
[the remark this answers is not in the document]
Comment (ry) [reply src/isolate.rs]: I tried it.. but it fails on 001_hello.js - which is a program that doesn't send any messages. So it blocks forever trying to receive something.

[the remark this answers is not in the document]
Comment (piscisaureus) [reply src/isolate.rs]: did you try:

```rs
    while !self.state.is_idle() {
          let box_u8 = self.rx.recv().unwrap();
          self.send(box_u8);
    }
```

[the remark this answers is not in the document]
Comment (ry) [reply src/isolate.rs]: Thanks - that works.

File: src/handlers.rs [hunk 2]
```diff
@@ -1,46 +1,44 @@
...
 type OpResult = DenoResult<Buf>;
 
 // TODO Ideally we wouldn't have to box the Op being returned.
 // The box is just to make it easier to get a prototype refactor working.
-type Handler = fn(i: *const isolate, base: &msg::Base) -> Box<Op>;
+type Handler = fn(state: Arc<IsolateState>, base: &msg::Base) -> Box<Op>;
 
-pub extern "C" fn msg_from_js(i: *const isolate, buf: deno_buf) {
-  let bytes = unsafe { std::slice::from_raw_parts(buf.data_ptr, buf.data_len) };
+// Hopefully Rust optimizes this away.
```
Comment (piscisaureus) [inline src/handlers.rs]: That seems a little unlikely. If you returned an &'static Buf it probably would be. But I don't think it's all that important.

File: src/isolate.rs [hunk 3]
```diff
@@ -1,54 +1,130 @@
...
+// Isolate cannot be passed between threads but IsolateState can. So any state that
+// needs to be accessed outside the main V8 thread should be inside IsolateState.
+pub struct IsolateState {
   pub dir: deno_dir::DenoDir,
-  pub rt: tokio::runtime::current_thread::Runtime,
-  pub timers: HashMap<u32, futures::sync::oneshot::Sender<()>>,
+  pub timers: Mutex<HashMap<u32, futures::sync::oneshot::Sender<()>>>,
   pub argv: Vec<String>,
   pub flags: flags::DenoFlags,
+  ntasks: Mutex<i32>,
```
Comment (piscisaureus) [inline src/isolate.rs]: I don't think the tasks count should is modified from any thread other than the main thread. 
(If this PR does that, I object to it).
So this mutex is unnecessary.

File: src/isolate.rs [hunk 4]
```diff
@@ -76,22 +170,70 @@ impl Drop for Isolate {
...
+
+  if is_sync {
+    // Execute op synchronously.
+    let buf = tokio_util::block_on(op).unwrap();
+    if buf.len() != 0 {
+      // Set the synchronous response, the value returned from isolate.send().
+      isolate.set_response(buf);
+    }
+  } else {
+    // Execute op asynchornously.
```
Comment (piscisaureus) [inline src/isolate.rs]: "asynchronously"

File: src/isolate.rs (hunk 3 above)
Comment (ry) [reply src/isolate.rs]: Fixed in ab50c33

File: src/isolate.rs (hunk 4 above)
Comment (ry) [reply src/isolate.rs]: Fixed.

File: src/handlers.rs (hunk 2 above)
Comment (ry) [reply src/handlers.rs]: Box is always `'static`. I doubt this is anything that will ever show up in a profile, I'd rather not complicate the function now without doing benchmarks.

**Comment by ry** [comment]
You should demand another test of Isolate dispatch : ) 
I will update with that soon.

**Pull request merged**